File: WoofAndroid/presentation/profile/src/main/kotlin/com/perrystreet/woof/presentation/profile/viewmodel/ProfileHeaderViewModel.kt
package com.perrystreet.woof.presentation.profile.viewmodel

import com.perrystreet.woof.models.dog.Dog
import com.perrystreet.woof.presentation.common.rx.UiObservable
import com.perrystreet.woof.presentation.common.rx.UiObservable.Companion.asUiObservable
import com.perrystreet.woof.presentation.common.viewmodel.StateDerivingViewModel
import com.perrystreet.woof.presentation.profile.uimodel.ProfileOverflowMenuItemUIModel
import com.perrystreet.woof.usecase.favorites.IsDogFavoriteUseCase
import com.perrystreet.woof.usecase.favorites.ToggleDogFavoriteUseCase
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.kotlin.plusAssign
import io.reactivex.rxjava3.subjects.BehaviorSubject
import org.koin.android.annotation.KoinViewModel
import org.koin.core.annotation.InjectedParam
import java.util.Optional

@KoinViewModel
class ProfileHeaderViewModel(
    @InjectedParam private val dog: Dog,
    private val toggleDogFavoriteUseCase: ToggleDogFavoriteUseCase,
    isDogFavoriteUseCase: IsDogFavoriteUseCase,
) : StateDerivingViewModel<ProfileHeaderViewModel.State>() {

    data class State(
        val name: String,
        val isFavorite: Boolean,
        val isOverflowExpanded: Boolean,
        val overflowItems: List<ProfileOverflowMenuItemUIModel>,
    )

    private val overflowExpanded: BehaviorSubject<Boolean> = BehaviorSubject.createDefault(false)

    override val state: UiObservable<State> = Observable.combineLatest(
        isDogFavoriteUseCase(dog),
        overflowExpanded,
    ) { isFavorite, isOverflowExpanded ->
        State(
            name = dog.name,
            isFavorite = isFavorite,
            isOverflowExpanded = isOverflowExpanded,
            overflowItems = ProfileOverflowMenuItemUIModel.entries,
        )
    }.asUiObservable(
        State(
            name = dog.name,
            isFavorite = false,
            isOverflowExpanded = false,
            overflowItems = ProfileOverflowMenuItemUIModel.entries,
        ),
    )

    fun onFavoriteTap() {
        disposables += toggleDogFavoriteUseCase(dog).subscribe(
            {},
            { error -> mutableError.onNext(Optional.of(error)) },
        )
    }

    fun onOverflowExpandedChange(isExpanded: Boolean) {
        overflowExpanded.onNext(isExpanded)
    }
}

File: WoofIOS/SwiftPackages/Presentation/Profile/Sources/PresentationProfile/ViewModel/ProfileHeaderViewModel.swift
import Combine
import DI
import Models
import PresentationCommon
import UseCase
import Utils

@Factory
public final class ProfileHeaderViewModel: StateDerivingViewModel<ProfileHeaderViewModel.State, FavoritesError> {
    @DIArgument private let dog: Dog
    private let toggleDogFavoriteUseCase: ToggleDogFavoriteUseCase
    private let overflowExpanded: CurrentValueSubject<Bool, Never>

    init(
        dog: Dog,
        toggleDogFavoriteUseCase: ToggleDogFavoriteUseCase,
        isDogFavoriteUseCase: IsDogFavoriteUseCase
    ) {
        let overflowExpanded = CurrentValueSubject<Bool, Never>(false)
        self.dog = dog
        self.toggleDogFavoriteUseCase = toggleDogFavoriteUseCase
        self.overflowExpanded = overflowExpanded
        super.init(
            initialValue: State(
                name: dog.name,
                isFavorite: false,
                isOverflowExpanded: false,
                overflowItems: ProfileOverflowMenuItemUIModel.allCases
            ),
            source: Publishers.CombineLatest(
                isDogFavoriteUseCase(dog: dog),
                overflowExpanded
            )
            .map { isFavorite, isOverflowExpanded in
                State(
                    name: dog.name,
                    isFavorite: isFavorite,
                    isOverflowExpanded: isOverflowExpanded,
                    overflowItems: ProfileOverflowMenuItemUIModel.allCases
                )
            }
            .eraseToAnyPublisher()
        )
    }

    public func onFavoriteTap() {
        toggleDogFavoriteUseCase(dog: dog)
            .pss_sink { [weak self] result in
                guard case .failure(let error) = result else { return }
                self?.emitError(error)
            }
            .store(in: &cancellables)
    }

    public func onOverflowExpandedChange(_ isExpanded: Bool) {
        overflowExpanded.send(isExpanded)
    }
}

extension ProfileHeaderViewModel {
    public struct State: Equatable {
        public let name: String
        public let isFavorite: Bool
        public let isOverflowExpanded: Bool
        public let overflowItems: [ProfileOverflowMenuItemUIModel]

        public init(name: String, isFavorite: Bool, isOverflowExpanded: Bool, overflowItems: [ProfileOverflowMenuItemUIModel]) {
            self.name = name
            self.isFavorite = isFavorite
            self.isOverflowExpanded = isOverflowExpanded
            self.overflowItems = overflowItems
        }
    }
}
